### DISCUSSION TITLE
How to create PSObject-wrapped value type

### DISCUSSION THREAD

**Discussion by ayeltsov** [PowerShell/PowerShell#15037] [Q&A]
For my unit tests I'd like to create a PSObject-wrapped value type.
Is there a good way of doing that?

P.S. Why: when Powershell wraps value types (like System.Int32) into PSObject, it can create trouble for interoperability with c# code which expects int but gets PSObject. 
I found this in scenario involving Invoke-Command on another machine. Looking for something simpler to add a unit test.

**Comment by powercode** [comment]
Short answer: You can't.

PSObject is by definition a class, and there is no language support in PowerShell for creating a ValueType, except for using `Add-Type` with inline C# code.

PSObject will box any value type you hand it, as it only takes deals with `object` in its BaseObject. 

PowerShell will on the other hand try to convert the boxed ValueType in its BaseObject when calling API:s.

Can you give a more specific example of when you are running into trouble? Perhaps we can provide other solutions?

**Reply by ayeltsov** [reply]
>Can you give a more specific example of when you are running into trouble?

The repro is rather convoluted and I'm not sure I want to spend time distilling it into a shorter one.

Overall description is: c# executes script and at the end retrieves the result:

object value = powerShell.Runspace.SessionStateProxy.PSVariable.GetValue("result", null);

In one case, script runs something like this, and c# sees that value as Int32 (or something that Convert.ToInt32 has no problem with):

```
Invoke-Command -ErrorAction Stop -VMName $vmMachineName -Credential $ConnectToVmCredentials -ScriptBlock `
                { 
                    Invoke-Expression $using:command | Tee-Object -Variable output
                    return [PSCustomObject]@{Output=$output; ExitCode=$LASTEXITCODE } 
                } | Tee-Object -Variable invokeCommandOutput
<...>
$result = $invokeCommandOutput.ExitCode
```

In another case, the code is like this, and Convert.ToInt32 fails with `System.InvalidCastException: Unable to cast object of type 'System.Management.Automation.PSObject' to type 'System.IConvertible'.
   at System.Convert.ToInt32(Object value)`

```
$scriptBlock = { invoke-expression 'cmd.exe /c exit 44'; return [PSCustomObject] @{ ExitCode=$LASTEXITCODE } }
<...>
                Invoke-Command -ErrorAction Stop -VMName $vmMachineName -Credential $ConnectToVmCredentials -ScriptBlock $scriptBlock `
                | Tee-Object -Variable invokeCommandOutput
<...>
        $result = $invokeCommandOutput | `
            where {$_.GetType() -eq [System.Management.Automation.PSCustomObject] -and $_.PSObject.Properties.Name -contains 'ExitCode'} | Select-Object -ExpandProperty ExitCode -Last 1

```

When I looked at the 2nd type in VS debugger, it was Int32 wrapped into PSObject, so as workaround I replaced Convert.ToInt32 with System.Management.Automation.LanguagePrimitives.ConvertTo<int> and it just worked.

I still did not fully understand what's going on. If I looked at the type in Powershell, it was showing `System.Int32, System.Private.CoreLib, Version=5.0.0.0` in both cases.

So... it sort of works right now, I'm not sure how much more time I want to spend on it.

**Reply by powercode** [reply]
LanguagePrimitive.ConvertTo tries many more ways of converting a PSObject to the result type than Convert.ToInt32 does.

But if you know that you expect an Int32, you should also be able to do something like

```CSharp
PSOBject psobj = ...;
if (psojb?.BaseObject is int i){
   // use your int.
   Assert.AreEqual(ExpectedValue, i);
}
else {
 // unexpected
  Assert.Fail("Unexpected return type: " + psob?.BaseObject?.GetType()); in your unit test
}
```

**Reply by ayeltsov** [reply]
Sure thing, thank you. 

My question was a bit more on how can I repro this situation where int ends up being wrapped into PSObject, but I'm not blocked at the moment and if I can't get this, it's not the end of the world.

BTW, this is how this PSObject looks in VS debugger:

```
? value.GetType()
{Name = "PSObject" FullName = "System.Management.Automation.PSObject"}
    Assembly: {System.Management.Automation, Version=7.1.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35}
    AssemblyQualifiedName: "System.Management.Automation.PSObject, System.Management.Automation, Version=7.1.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35"
    Attributes: Public | Serializable | BeforeFieldInit
    BaseType: {Name = "Object" FullName = "System.Object"}
    Cache: {System.RuntimeType.RuntimeTypeCache}
    CacheIfExists: {System.RuntimeType.RuntimeTypeCache}
    ContainsGenericParameters: false
    CustomAttributes: Count = 0x00000002
    DeclaredConstructors: {System.Reflection.ConstructorInfo[0x00000005]}
    DeclaredEvents: {System.Reflection.EventInfo[0x00000000]}
    DeclaredFields: {System.Reflection.FieldInfo[0x00000029]}
    DeclaredMembers: {System.Reflection.MemberInfo[0x000000ba]}
    DeclaredMethods: {System.Reflection.MethodInfo[0x0000006c]}
    DeclaredNestedTypes: {System.Reflection.TypeInfo.<get_DeclaredNestedTypes>d__22}
    DeclaredProperties: {System.Reflection.PropertyInfo[0x00000019]}
    DeclaringMethod: '((System.RuntimeType)value.GetType()).DeclaringMethod' threw an exception of type 'System.InvalidOperationException'
    DeclaringType: null
    DomainInitialized: false
    FullName: "System.Management.Automation.PSObject"
    GUID: {388c0d35-9adf-3bc7-90de-f9adf60fb761}
    GenericCache: null
    GenericParameterAttributes: '((System.RuntimeType)value.GetType()).GenericParameterAttributes' threw an exception of type 'System.InvalidOperationException'
    GenericParameterPosition: '((System.RuntimeType)value.GetType()).GenericParameterPosition' threw an exception of type 'System.InvalidOperationException'
    GenericTypeArguments: {System.Type[0x00000000]}
    GenericTypeParameters: {System.Type[0x00000000]}
    HasElementType: false
    ImplementedInterfaces: {System.Type[0x00000004]}
    IsAbstract: false
    IsAnsiClass: true
    IsArray: false
    IsAutoClass: false
    IsAutoLayout: true
    IsByRef: false
    IsByRefLike: false
    IsCOMObject: false
    IsClass: true
    IsCollectible: false
    IsConstructedGenericType: false
    IsContextful: false
    IsEnum: false
    IsExplicitLayout: false
    IsGenericMethodParameter: false
    IsGenericParameter: false
    IsGenericType: false
    IsGenericTypeDefinition: false
    IsGenericTypeParameter: false
    IsImport: false
    IsInterface: false
    IsLayoutSequential: false
    IsMarshalByRef: false
    IsNested: false
    IsNestedAssembly: false
    IsNestedFamANDAssem: false
    IsNestedFamORAssem: false
    IsNestedFamily: false
    IsNestedPrivate: false
    IsNestedPublic: false
    IsNotPublic: false
    IsPointer: false
    IsPrimitive: false
    IsPublic: true
    IsSZArray: false
    IsSealed: false
    IsSecurityCritical: true
    IsSecuritySafeCritical: false
    IsSecurityTransparent: false
    IsSerializable: true
    IsSignatureType: false
    IsSpecialName: false
    IsTypeDefinition: true
    IsUnicodeClass: false
    IsValueType: false
    IsVariableBoundArray: false
    IsVisible: true
    MemberType: TypeInfo
    MetadataToken: 0x02000329
    Module (System.Reflection.MemberInfo): {System.Management.Automation.dll}
    Module: {System.Management.Automation.dll}
    Name: "PSObject"
    Namespace: "System.Management.Automation"
    ReflectedType: null
    StructLayoutAttribute: {System.Runtime.InteropServices.StructLayoutAttribute}
    TypeHandle: {System.RuntimeTypeHandle}
    TypeInitializer: {Void .cctor()}
    UnderlyingSystemType: {Name = "PSObject" FullName = "System.Management.Automation.PSObject"}
    m_cache: 0x000001ce8b3878f8
    m_handle: 0x00007ffe3c9fdd20
    m_keepalive: null

? ((System.Management.Automation.PSObject)value).BaseObject.GetType()
{Name = "Int32" FullName = "System.Int32"}
    Assembly: {System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e}
    AssemblyQualifiedName: "System.Int32, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
<...>
```

**Reply by powercode** [reply]
Any time you assign a variable in PowerShell, it will be a PSObject, so just doing
```powershell
$i = 10
```
will do that.

**Reply by powercode** [reply]
PSObject is the root of the object abstraction in PowerShell - the place where the dynamic nature of the language is hooked up.

But PSObject only knows about `object` so you then get the standard .net boxing for all value types on top of that.

**Reply by ayeltsov** [reply]
So something interesting must be happening when c# asks for the variable. 

In one case, it gets unboxed Int32, and in another case it gets PSObject. And if I do GetType() on powershell side, they are both showing Int32

Just double-checked - in the other case it's straight up Int32

```
            object value = powerShell.Runspace.SessionStateProxy.PSVariable.GetValue("result", null);

            return value;

? value.GetType()
{Name = "Int32" FullName = "System.Int32"}
    Assembly: {System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e}
    AssemblyQualifiedName: "System.Int32, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
```

**Reply by powercode** [reply]
https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/MshObject.cs

PSObject will try to delegate method calls in powershell to the wrapped object.
If you want to actually access the abstraction, you can do 

```powershell
$i = 10
$i.psobject.GetType()

```

**Reply by SeeminglyScience** [reply]
> For my unit tests I'd like to create a PSObject-wrapped value type.
> Is there a good way of doing that?

`PSObject.AsPSObject(object)` or `new PSObject(object)`

But yeah, you always need to be prepared to unwrap a `PSObject` when dealing with PowerShell.  Sometimes it's wrapped, sometimes it's not.  It's good to always check because the specifics whether the result is wrapped is not (afaik) part of the change contract.
